lib/puppet/parser/compiler.rb

```diff
@@ -865,11 +865,32 @@ def create_settings_scope
     scope = @topscope.class_scope(settings_type)
 
     env = environment
+    settings_hash = {}
     Puppet.settings.each do |name, setting|
       next if name == :name
-      scope[name.to_s] = env[name]
+      name = name.to_s
+      # Construct a hash (in anticipation it will be set in top scope under a name like $settings)
+      settings_hash[name] = transform_setting(env[name])
+      scope[name.to_s] = settings_hash[name]
+    end
+  end
+
+  def transform_setting(val)
+    case val
+    when Integer, Float, String
+      val
+    when Symbol
+      val == :undef ? nil : val.to_s
+    when Array
+      val.map {|entry| transform_setting(entry) }
+    when Hash
+      Hash[val.map {|k,v| [transform_setting(k), transform_setting(v)] }]
```

Comment:
it is done once per compilation - so not time critical, still worth making it faster and produce less garbage.